phpBB/phpbb/ucp/controller/reset_password.php

```diff
@@ -0,0 +1,436 @@
+		$user_row = $this->db->sql_fetchrow($result);
+		$this->db->sql_freeresult($result);
+
+		$message = $this->language->lang('RESET_TOKEN_EXPIRED_OR_INVALID') . '<br /><br />' . $this->language->lang('RETURN_INDEX', '<a href="' . append_sid("{$this->root_path}index.{$this->php_ext}") . '">', '</a>');
+
+		if (empty($user_row))
+		{
+			return $this->helper->message($message);
+		}
+
+		if (!hash_equals($reset_token, $user_row['reset_token']))
+		{
+			return $this->helper->message($message);
+		}
+
+		if ($user_row['reset_token_expiration'] < time())
+		{
+			$this->remove_reset_token($user_id);
+
+			return $this->helper->message($message);
+		}
+
+		$error = [];
+
+		if ($submit)
+		{
+			if (!check_form_key('ucp_reset_password'))
+			{
+				return $this->helper->message('FORM_INVALID');
+			}
+
+			if ($user_row['user_type'] == USER_IGNORE || $user_row['user_type'] == USER_INACTIVE)
+			{
+				return $this->helper->message($message);
+			}
+
+			// Check users permissions
+			$auth2 = new auth();
+			$auth2->acl($user_row);
+
+			if (!$auth2->acl_get('u_chgpasswd'))
+			{
+				return $this->helper->message($message);
+			}
+
+			if (!function_exists('validate_data'))
+			{
+				include($this->root_path . 'includes/functions_user.' . $this->php_ext);
+			}
+
+			$data = [
+				'new_password'		=> $this->request->untrimmed_variable('new_password', '', true),
+				'password_confirm'	=> $this->request->untrimmed_variable('new_password_confirm', '', true),
+			];
+			$check_data = [
+				'new_password'		=> [
+					['string', false, $this->config['min_pass_chars'], $this->config['max_pass_chars']],
+					['password'],
+				],
+				'password_confirm'	=> ['string', true, $this->config['min_pass_chars'], $this->config['max_pass_chars']],
+			];
+			$error = array_merge($error, validate_data($data, $check_data));
+			if (strcmp($data['new_password'], $data['password_confirm']) !== 0)
+			{
+				$error[] = ($data['password_confirm']) ? 'NEW_PASSWORD_ERROR' : 'NEW_PASSWORD_CONFIRM_EMPTY';
+			}
+			if (empty($error))
+			{
+				$sql_ary = [
+					'user_password'				=> $this->passwords_manager->hash($data['new_password']),
+					'user_login_attempts'		=> 0,
+					'reset_token'				=> '',
+					'reset_token_expiration'	=> 0,
+				];
+				$sql = 'UPDATE ' . $this->tables['users'] . '
+							SET ' . $this->db->sql_build_array('UPDATE', $sql_ary) . '
+							WHERE user_id = ' . (int) $user_row['user_id'];
+				$this->db->sql_query($sql);
+				$this->log->add('user', $user_row['user_id'], $this->user->ip, 'LOG_USER_NEW_PASSWORD', false, [
+					'reportee_id' => $user_row['user_id'],
+					$user_row['username']
+				]);
+				meta_refresh(3, append_sid("{$this->root_path}index.{$this->php_ext}"));
+				return $this->helper->message($this->language->lang('PASSWORD_RESET'));
+			}
+		}
+
+		$this->template->assign_vars([
+			'S_IS_PASSWORD_RESET'	=> true,
+			'ERROR'					=> !empty($error) ? implode('<br />', array_map([$this->language, 'lang'], $error)) : '',
```

Comment:
I think you'll make Hanakin happy if we stop doing this in PHP.
We can just assign the array to the template and do it in twig.
```html
{% if ERROR %}
    {% for error in ERROR %}{{ error }}<br>{% endfor %}
{% endif %}
```
So eventually he can do all the styling and stuff in HTML / CSS.